extensions/src/main/java/org/pentaho/platform/web/http/api/resources/services/ActionService.java

```diff
@@ -0,0 +1,180 @@
+/*!
+ * This program is free software; you can redistribute it and/or modify it under the
+ * terms of the GNU Lesser General Public License, version 2.1 as published by the Free Software
+ * Foundation.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along with this
+ * program; if not, you can obtain a copy at http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
+ * or from the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Lesser General Public License for more details.
+ *
+ * Copyright (c) 2017 Pentaho Corporation..  All rights reserved.
+ */
+
+package org.pentaho.platform.web.http.api.resources.services;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.http.HttpStatus;
+import org.pentaho.platform.action.ActionInvokeStatus;
+import org.pentaho.platform.api.action.IAction;
+import org.pentaho.platform.api.action.IActionInvokeStatus;
+import org.pentaho.platform.api.action.IActionInvoker;
+import org.pentaho.platform.engine.core.system.PentahoSessionHolder;
+import org.pentaho.platform.engine.core.system.PentahoSystem;
+import org.pentaho.platform.plugin.action.ActionParams;
+import org.pentaho.platform.plugin.action.LocalActionInvoker;
+import org.pentaho.platform.plugin.action.messages.Messages;
+import org.pentaho.platform.util.ActionUtil;
+import org.pentaho.platform.util.StringUtil;
+import org.pentaho.platform.web.http.api.resources.WorkerNodeActionInvokerAuditor;
+import org.pentaho.platform.workitem.WorkItemLifecyclePhase;
+import org.pentaho.platform.workitem.WorkItemLifecycleEventUtil;
+import org.slf4j.MDC;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * This service performs action related tasks, such as running/invoking the action in the background.
+ */
+public class ActionService {
+
+    protected static final Log logger = LogFactory.getLog( ActionService.class );
+    protected static final int MAX_THREADS = 8;
+    protected static ExecutorService executorService = Executors.newFixedThreadPool( MAX_THREADS,
+            new ThreadFactoryBuilder().setNameFormat( "worker-thread-%d" ).build() );
```

Comment:
Please make sure the factory builds daemon threads. (I don't remember if they are marked as daemon threads when using those methods)